facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+    // was set up with, if one is not already held or it has expired. Otherwise
+    // reutrns a copy of the valid token.
+    fn ensure_impersonated_service_account_oauth_token(&mut self) -> Result<String> {
+        if self.service_account_to_impersonate.is_none() {
+            return Err(anyhow!("no service account to impersonate was provided"));
+        }
+        let service_account_to_impersonate = self.service_account_to_impersonate.clone().unwrap();
+        // API reference:
+        // https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
+        let request_url = format!("https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/{}:generateAccessToken",
+            service_account_to_impersonate);
+        let http_response = ureq::post(&request_url)
+            .set(
+                "Authorization",
+                &format!(
+                    "Bearer {}",
+                    self.ensure_default_service_account_oauth_token()?
+                ),
+            )
+            .set("Content-Type", "application/json")
+            // At the moment, these tokens are only used to read and write from
+            // GCS buckets, so we request an appropriate scope. We could further
+            // narrow this to a devstorage.read token if we knew what it was
+            // going to be used for, but that would require this object to
+            // manage two impersonation tokens which is more work than I want to
+            // do right now.
+            // https://cloud.google.com/storage/docs/authentication#oauth-scopes
+            .send_json(ureq::json!({
+                "scope": [
+                    "https://www.googleapis.com/auth/devstorage.read_write"
+                ]
+            }));
+        if http_response.error() {
+            return Err(anyhow!(
+                "failed to get Oauth token to impersonate service account {}: {:?}",
+                service_account_to_impersonate,
+                http_response
+            ));
+        }
+
+        let response = http_response
+            .into_json_deserialize::<GenerateAccessTokenResponse>()
+            .context("failed to deserialize response from IAM API")?;
+        self.impersonated_service_account_oauth_token = Some(OauthToken {
+            token: response.access_token.clone(),
+            expiration: response.expire_time,
+        });
+
+        Ok(response.access_token)
+    }
+}
+
+// GCSTransport manages reading and writing from GCS buckets. Currently, the
+// only authentication it supports is by means of impersonating a specified GCP
+// service account, which is assumed to have the necessary read or write
+// permissions on the bucket.
+pub struct GCSTransport {
+    path: GCSPath,
+    oauth_token_provider: OauthTokenProvider,
+}
+
+impl GCSTransport {
+    pub fn new(path: GCSPath, impersonate: Option<String>) -> GCSTransport {
```

Comment:
This needs documentation for what the value of `impersonate` should be. Also, it looks like it's an error to pass None here; do I understand correctly? Maybe we should make this a `String` if so.

Reply:
It's not optional, because we also support authenticating as the default service account (the one whose credentials are available in the GKE metadata service). We support the impersonation case so that our per-data share processor GCP service accounts can impersonate the singleton sum part bucket writer service account and write sum parts to the MITRE owned buckets. We support the default SA case so that once our own storage moves to GCS, we can bless our per-data share processor SAs to use the buckets we create. I put in a doccomment.